### ISSUE TITLE
New rule proposal: prefer-regex-literals

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12238]
**Please describe what the rule should do:**

Suggests to use regex literals instead of `RegExp()`/`new RegExp()` when these calls have:

* One argument that is a string literal.
* Two arguments and they are both string literals.

**What category of rule is this? (place an "X" next to just one item)**

[X] Suggests an alternate way of doing something (suggestion)

**Provide 2-3 code examples that this rule will warn about:**

```js
// good, dynamic
new RegExp(pattern);
new RegExp(prefix + "foo");
new RegExp("foo", flags);

// good, regex literals
/^\d\./
/\d\d\.\d\d\.\d\d\d\d/

// bad, this is a very possible error
new RegExp("^\d\.");
new RegExp("\d\d\.\d\d\.\d\d\d\d");

// bad, correct but unreadable
new RegExp("^\\d\\.");
new RegExp("\\d\\d\\.\\d\\d\\.\\d\\d\\d\\d");
```

**Why should this rule be included in ESLint (instead of a plugin)?**

I think this is not just a style preference, rather a good practice to avoid an unnecessary level of the string literal syntax on top of the regex syntax. It's error-prone and difficult to reason about escapes of escapes.

**Are you willing to submit a pull request to implement this rule?**

Yes.

**Comment by platinumazure** [comment]
Yes! I really want to see this in core. :+1: from me.

**Comment by mysticatea** [comment]
How about ``new RegExp(String.raw`\d\d\.\d\d\.\d\d\d\d`);``?

**Comment by mdjermanovic** [comment]
> How about `` new RegExp(String.raw`\d\d\.\d\d\.\d\d\d\d`); ``?

That style indeed doesn't have the problem with double escapes.

First idea was to disallow just string literals (maybe also simple template literals used as strings), but since the rule is `prefer-regex-literals` it might make sense to disallow this style, too?

**Comment by mysticatea** [comment]
> since the rule is prefer-regex-literals it might make sense to disallow this style, too?

Yes, I think so.

**Comment by mdjermanovic** [comment]
I agree! Working on this.
